lib/Modem.ts

```diff
@@ -0,0 +1,829 @@
+		const messageID = `${Date.now()}`;
+		const submit = new pdu.Submit(number, message);
+
+		if (alert) {
+			submit.dataCodingScheme.setUseMessageClass(true);
+		}
+
+		// An SMS has a character limit, which is why the message is divided into several parts
+		const parts = submit.getParts();
+
+		const cmdSequence: prTypes.CmdSequence = {
+			cmds: [],
+			cancelOnFailure: true
+		};
+
+		const checkReponse = (response: CommandResponse | Error) => {
+			if (
+				response instanceof Error ||
+				response[response.length].toUpperCase().includes('ERROR') ||
+				!(response[response.length].trim() === '>' || response[response.length].trim().toUpperCase() === 'OK')
+			) {
+				throw new Error(`serialport-gsm/${this.port.path}: Failed to send SMS!`);
+			}
+
+			console.log('Reponse: ', response);
+		};
+
+		for (const part of parts) {
+			cmdSequence.cmds.push(
+				...[
+					new Command(`AT+CMGS=${part.toString(submit).length / 2 - 1}`, 1000),
+					new Command(`${part.toString(submit)}` + '\x1a', undefined, checkReponse)
+				]
+			);
+		}
+
+		// Sending all parts
+		await new Promise((resolve: (success: true) => void, reject: (error: Error) => void) => {
+			cmdSequence.onFinish = () => resolve(true);
+
+			cmdSequence.onFailed = (error) => {
+				const result: SendSMSFailed = {
+					success: false,
+					messageID,
+					error
+				};
+
+				this.events.emitMessageSendingFailed(result);
+				reject(error);
+			};
+
+			this.pushToQue(cmdSequence, prio);
+		});
+
+		const result: SendSMSSuccess = {
+			success: true,
+			messageID,
+			data: {
+				message,
+				recipient: number,
+				alert,
+				pdu: submit
+			}
+		};
+
+		this.events.emitSMSsent(result);
+		return result;
+	}
+
+	async getSignalInfo(prio = false) {
+		const response = await this.executeATCommand('AT+CSQ', true, prio);
+
+		if (!response.startsWith('+CSQ:') || response.toUpperCase().includes('ERROR')) {
+			throw new Error(`serialport-gsm/${this.port.path}: The network signal could not be read!`);
+		}
+
+		if (!response.split(' ')[1] || !response.split(' ')[1].split(',')[0]) {
+			throw new Error(`serialport-gsm/${this.port.path}: The signal strength could not be parsed!`);
+		}
+
+		const signal = Number(response.split(' ')[1].split(',')[0]);
+
+		if (isNaN(signal) || signal < 0 || signal > 31) {
+			throw new Error(`serialport-gsm/${this.port.path}: The parsed signal strength is invalid!`);
+		}
+
+		return {
+			signalQuality: signal,
+			signalStrength: 113 - signal * 2
+		};
+	}
+
+	async checkSimMemory(prio = false) {
+		const response = await this.executeATCommand('AT+CPMS="SM"', true, prio);
+
+		if (!response.startsWith('+CME') || response.toUpperCase().includes('ERROR')) {
+			throw new Error(`serialport-gsm/${this.port.path}: The required memory space of the SMS, could not be read!`);
+		}
+
+		const s = response.split(' ')[1];
+
+		if (!s || isNaN(Number(s.split(',')[0])) || isNaN(Number(s.split(',')[1]))) {
```

Comment:
We should hnot split that string 4 times in 3 rows :-)

Reply:
Is hopefully improved in the next update